### ISSUE TITLE
getting invalid system call

### ISSUE THREAD

**Issue by Mohamed-Fathy-Salah** [stoeckmann/xwallpaper#42]
OS : arch linux
xwallpaper version : 0.7.3

after i run `xwallpaper --zoom wall.jpg` i get the following error message `invalid system call (core dumped)`

any idea how to fix it?

**Comment by stoeckmann** [comment]
I have tested with Arch Linux myself (amd64) but was unable to reproduce this. Are you using the testing branch? Or another platform?

Anyway, please run xwallpaper with strace, e.g. `strace xwallpaper --zoom wall.jpg` and post the last lines (or everything if in doubt). Somewhere at the end a system call should fail. This means that the sandbox of xwallpaper is too tight and something has changed in one of the underlying libraries.

The strace output should look like the one in issue https://github.com/stoeckmann/xwallpaper/issues/35.

If you are able to compile xwallpaper on your system, then you might add the system calls to https://github.com/stoeckmann/xwallpaper/blob/master/seccomp.c#L129 (try stage1 first and stage2 if it still fails). Otherwise I'll do that as soon as I know which call fails.

**Comment by Mohamed-Fathy-Salah** [comment]
I am using the app in pacman package manager.

output of `strace xwallpaper --zoom wall.jpg`

```
prctl(PR_CAPBSET_READ, CAP_MAC_OVERRIDE) = 1
prctl(PR_CAPBSET_READ, 0x30 /* CAP_??? */) = -1 EINVAL (Invalid argument)
prctl(PR_CAPBSET_READ, CAP_CHECKPOINT_RESTORE) = 1
prctl(PR_CAPBSET_READ, 0x2c /* CAP_??? */) = -1 EINVAL (Invalid argument)
prctl(PR_CAPBSET_READ, 0x2a /* CAP_??? */) = -1 EINVAL (Invalid argument)
prctl(PR_CAPBSET_READ, 0x29 /* CAP_??? */) = -1 EINVAL (Invalid argument)
prctl(PR_GET_SECCOMP)                   = 0
prctl(PR_SET_SECCOMP, SECCOMP_MODE_FILTER, NULL) = -1 EFAULT (Bad address)
seccomp(SECCOMP_SET_MODE_STRICT, 0x1, NULL) = -1 EINVAL (Invalid argument)
seccomp(SECCOMP_SET_MODE_FILTER, SECCOMP_FILTER_FLAG_TSYNC, NULL) = -1 EFAULT (Bad address)
seccomp(SECCOMP_SET_MODE_FILTER, SECCOMP_FILTER_FLAG_LOG, NULL) = -1 EFAULT (Bad address)
seccomp(SECCOMP_GET_ACTION_AVAIL, 0, [SECCOMP_RET_LOG]) = 0
seccomp(SECCOMP_GET_ACTION_AVAIL, 0, [SECCOMP_RET_KILL_PROCESS]) = 0
seccomp(SECCOMP_SET_MODE_FILTER, SECCOMP_FILTER_FLAG_SPEC_ALLOW, NULL) = -1 EFAULT (Bad address)
seccomp(SECCOMP_SET_MODE_FILTER, SECCOMP_FILTER_FLAG_NEW_LISTENER, NULL) = -1 EFAULT (Bad address)
seccomp(SECCOMP_GET_NOTIF_SIZES, 0, {seccomp_notif=80, seccomp_notif_resp=24, seccomp_data=64}) = 0
seccomp(SECCOMP_SET_MODE_FILTER, SECCOMP_FILTER_FLAG_TSYNC_ESRCH, NULL) = -1 EFAULT (Bad address)
prctl(PR_SET_NO_NEW_PRIVS, 1, 0, 0, 0)  = 0
seccomp(SECCOMP_SET_MODE_FILTER, 0, {len=95, filter=0x55a093ae4ed0}) = 0
openat(AT_FDCWD, ".config/wall.jpg", O_RDONLY) = 3
newfstatat(3, "", {st_mode=S_IFREG|0644, st_size=1416286, ...}, AT_EMPTY_PATH) = 0
socket(AF_UNIX, SOCK_STREAM|SOCK_CLOEXEC, 0) = 4
connect(4, {sa_family=AF_UNIX, sun_path=@"/tmp/.X11-unix/X0"}, 20) = 0
getpeername(4, {sa_family=AF_UNIX, sun_path=@"/tmp/.X11-unix/X0"}, [124 => 20]) = 0
uname({sysname="Linux", nodename="arch", ...}) = 0
access("/home/mofasa/.Xauthority", R_OK) = 0
openat(AT_FDCWD, "/home/mofasa/.Xauthority", O_RDONLY) = 5
newfstatat(5, "", {st_mode=S_IFREG|0600, st_size=49, ...}, AT_EMPTY_PATH) = 0
read(5, "\1\0\0\4arch\0\0010\0\22MIT-MAGIC-COOKIE-1\0"..., 4096) = 49
read(5, "", 4096)                       = 0
futex(0x7f0cca352094, FUTEX_WAKE_PRIVATE, 2147483647) = 202
+++ killed by SIGSYS (core dumped) +++

```

I am pretty sure that this error happened after the last upgrade, but I am not sure what package caused it.

**Comment by stoeckmann** [comment]
I have created a branch which has the `futex` systemcall in its sandbox. It would be great if you could test it.
In order to compile xwallpaper on Arch Linux, this should do the trick:
```
# Run this if you lack compiler and other build tools
sudo pacman -S base-devel git

git clone --branch futex https://github.com/stoeckmann/xwallpaper.git
cd xwallpaper
./autogen.sh
./configure
./xwallpaper --zoom <path_to_your_file>
```

When I run xwallpaper on Arch Linux like you do, the read calls for Xauthority are followed by a close. I assume that the code in question is within libXau but I am not sure what exactly triggers internally a futex call. Maybe some resolver which I do not have on my system...

Did you change something with DNS or xauth maybe?

Long story short: If futex is all that is needed, then we are good to go and I would release a new version.

**Comment by stoeckmann** [comment]
If this still fails, then you can also run `./configure --without-seccomp` instead, which disables the sandbox. If the program runs properly, then we have to find more blocked system calls. You could send me the whole strace output in this case to tobias@stoeckmann.org (or as a reply in this issue).

But if xwallpaper fails then as well, then there is something else failing. If it happens, you can run `coredumpctl info` and post the stack trace. Maybe we get more information out of it.

**Comment by Mohamed-Fathy-Salah** [comment]
I've tried the above commands but sadly it didn't work. however, when I tried compiling the code on that branch it worked (i.e. `make clean install` ).

**Comment by stoeckmann** [comment]
Oh yes, I totally forgot the make call. Sorry and thanks a lot for testing it. I'll release a new version so distributions can pick up this change!
